channel/recv.go

```diff
@@ -6,21 +6,22 @@ package channel
 
 import "context"
 
-// RecvWithContext tries to receive a value from a channel which is aborted if the context is canceled.
+// RecvWithContext tries to receive a value from a channel. Operation is aborted if the context is canceled.
 //
-// Function returns true if the value was received, false if the context was canceled or the channel was closed.
-func RecvWithContext[T any](ctx context.Context, ch <-chan T) (T, bool) {
+// Function returns [StateRecv] if the value was received, [StateCancelled] on context cancelation and
+// [StateClosed] if the channel was closed.
```

Comment:
do we even need this function in the long run?

Reply:
Yes. We [have](https://github.com/cosi-project/runtime/blob/bf3352ff24179649c8921b899468cc69b66567d7/pkg/controller/runtime/runtime.go#L290) [several](https://github.com/cosi-project/runtime/blob/bf3352ff24179649c8921b899468cc69b66567d7/pkg/controller/runtime/runtime.go#L350) [places](https://github.com/cosi-project/runtime/blob/bf3352ff24179649c8921b899468cc69b66567d7/pkg/controller/runtime/runtime.go#L188) [just](https://github.com/cosi-project/runtime/blob/bf3352ff24179649c8921b899468cc69b66567d7/pkg/resource/rtestutils/assertions.go#L77) in [runtime](https://github.com/cosi-project/runtime/blob/bf3352ff24179649c8921b899468cc69b66567d7/pkg/resource/rtestutils/assertions.go#L112) [where](https://github.com/cosi-project/runtime/blob/bf3352ff24179649c8921b899468cc69b66567d7/pkg/resource/rtestutils/destroy.go#L45) it is [fully](https://github.com/cosi-project/runtime/blob/bf3352ff24179649c8921b899468cc69b66567d7/pkg/safe/state.go#L165) [used](https://github.com/cosi-project/runtime/blob/bf3352ff24179649c8921b899468cc69b66567d7/pkg/state/wrap.go#L88) with all results. I'm thinking about shorting it to just `channel.Recv` tho.

Reply:
where it was used it looks like it's fine to keep the old way?

Reply:
I'm not sure. I don't like long blocks of code inside `case v := <-chanData:` statement. For places I showed it will mean one more indentation and its harder to process visually where such block ends. We could partially solve it using anonymous functions, but that doesn't look like a proper solution too.

Reply:
We could also accept some kind of convention, like:

```go
var val T
select {
case <-ctx.Done():
	return nil
case val <- chanData:
}
```

Which will remove the need for `RecvWithContext`